### ISSUE TITLE
ONNX BERTModel produces different number of outputs  between Python and C#

### ISSUE THREAD

**Issue by abgoswam** [microsoft/onnxruntime#3141]
**Describe the bug**
[BERTModel ](https://huggingface.co/transformers/model_doc/bert.html#bertmodel) with `config.output_hidden_states=True`  produces different number of outputs  

- Python OnnxRuntime produces 15 output tensors
- C# OnnxRuntime produces 14 output tensors

**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Windows 10
- ONNX Runtime installed from (source or binary): binary
- ONNX Runtime version: 1.1.2
- Python version:3.7
- Visual Studio version (if applicable):16.4.5

**To Reproduce**
Describe steps/code to reproduce the behavior:

- Create [BERTModel ](https://huggingface.co/transformers/model_doc/bert.html#bertmodel) with `config.output_hidden_states=True`
- Convert it to ONNX using `torch.onnx.export`  (torch version 1.4.0)
- Produce output in Python, see the number of tensors in the output (15)
- Produce output in C#, see the number of tensors in the output (14)

**Expected behavior**
The tensors should match in the count and their respective values

**Comment by abgoswam** [comment]
Attaching a snapshot

![image](https://user-images.githubusercontent.com/8822956/75936035-e4048300-5e35-11ea-92a6-525145167088.png)

Python has 2 outputs corresponding to "1609".  But C# has just 1 output entry from 1609

@yufenglee

**Comment by yufenglee** [comment]
@abgoswam , as we synced, this model is weird with duplicated outputs 1609, but onnx spec doesn't specify it is invalid. 
The reason C# has one less output is because C# uses a dictionary for the output meta: [https://github.com/microsoft/onnxruntime/blob/6a0ee7eff6ce42935c73e52f88fb6d48eb1ea53c/csharp/src/Microsoft.ML.OnnxRuntime/InferenceSession.cs#L246](url)

**Comment by abgoswam** [comment]
@yufenglee. My main concern is that for the _same_ ONNX model, Py and C# OnnxRuntime produce _different_ results.  This is inconsistent.  At the very minimum, I would expect OnnxRuntime to produce exact  outputs using OnnxRuntime Python and C# libraries.

Also, I do not think it's fair to say the model is weird.  

The [BERTModel ](https://huggingface.co/transformers/model_doc/bert.html#bertmodel) returns  output tensors.   Giving names to the output tensors / using a dictionary for metadata, etc are pure ONNX abstractions.  As an end user, the expectation is that ONNX should produce the same output tensors as PyTorch.

**Comment by jywu-msft** [comment]
> @abgoswam , as we synced, this model is weird with duplicated outputs 1609, but onnx spec doesn't specify it is invalid.
> The reason C# has one less output is because C# uses a dictionary for the output meta:
> 
> https://github.com/microsoft/onnxruntime/blob/6a0ee7eff6ce42935c73e52f88fb6d48eb1ea53c/csharp/src/Microsoft.ML.OnnxRuntime/InferenceSession.cs#L246

from https://github.com/onnx/onnx/blob/master/docs/IR.md
"The graph MUST adhere to single static assignment (SSA) for all node outputs; this means that all node output names MUST be unique within a graph."
see example of such an issue in the early days of onnx: https://github.com/onnx/models/issues/2

**Comment by yufenglee** [comment]
Thanks @jywu-msft ! This issue is different from the onnx/models#2. Here the output name of node Add is unique:1609, but it is added twice to the output of the graph. We discussed this issue, and think it is not reasonable to have duplicate output for graph. ONNX will add a check for this.

> The BERTModel returns output tensors. Giving names to the output tensors / using a dictionary for metadata, etc are pure ONNX abstractions. As an end user, the expectation is that ONNX should produce the same output tensors as PyTorch.

The output hidden_states of [BERTModel](https://huggingface.co/transformers/model_doc/bert.html#bertmodel) is one tensor in PyTorch, but splitted into 12 different tensors when exporting. This is an issue in PyTorch exporter. They will fix it. 

To summarize the issue, 
PyTorch ONNX exporter will fix the duplicated output name issue.
ONNX will add a check that there is no duplicate input and output names
ONNXRuntime will add a check and fail early if model has duplicate input or output names.

**Comment by hariharans29** [comment]
I think OnnxRuntime can't add that check to fail early as adding this check could break existing (invalid) models being used via the Python API (which has no issues in case there are dupes). As you said, the onnx checker can be enhanced to add this.
